### ISSUE TITLE
Unexpected behavior for `Subarray` ranges

### ISSUE THREAD

**Issue by fgdorais** [leanprover/lean4#2360]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

This was [discussed on Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Unusual.20Subarray.20behavior/near/356388728) a while ago but I forgot to submit a bug report.

```lean
def a : Array Nat := #[0,1,2,3,4,5,6]

def b : Subarray Nat := a[1:5]

#eval (b.as, b.start, b.stop) -- (#[0,1,2,3,4,5,6], 1, 5)

def c : Subarray Nat := b[1:3]

#eval (c.as, c.start, c.stop) -- (#[1,2,3,4], 1, 3) !?
```

The expected result for the last eval is `(#[0,1,2,3,4,5,6], 2, 4)`.

The reason for this odd behavior is that the [`a[start:stop]` notation](https://github.com/leanprover/lean4/blob/53477089fe255280d7a816371ec315c6e2e94220/src/Init/Data/Array/Subarray.lean#L154-L161) assumes that `a` is an array. So when calculating `c`, Lean first coerces `b` into an array and then calculates the subarray of that.

### Versions

`Lean (version 4.0.0-nightly-2023-07-15, commit d37bbf4292c7, Release)`

### Additional Information

In my opinion, there should not be a coercion from `Subarray` to `Array` (which involves a possibly expensive array copy), but there should be one from `Array` to `Subarray` (which is essentially zero cost). Then the `a[start:stop]` could be refactored to assume that `a` is a subarray instead.

The `Subarray` to `Array` coercion is not used in `Init` in any essential way, but it is used lot in `Lean` so this would not be a trivial change.

**Comment by semorrison** [comment]
I'm going to label as `low-priority` for now. I think *if* removing the coercion from `Subarray` to `Array` could be done without making existing code harder to follow **and** it resulted in a concrete performance improvement somewhere, then it would be good to consider.

I agree that the potentially expensive yet invisible coercion could be a footgun.

**Comment by fgdorais** [comment]
@semorrison I agree, everything is working fine right now so there is no pressing need to fix this.

Note that removing the coercion from `Subarray` to `Array` wouldn't fix the unexpected behavior by itself, it merely explains the behavior. So, just for clarification purposes, does `low-priority` apply to the main issue presented here, the unusual coercion, or both. I'm assuming it's both but I would appreciate the clarification.

**Comment by semorrison** [comment]
Sorry for the slow reply. Yes to both.

I suspect that `Subarray` was simply not rewritten with uses outside the compiler in mind, and we don't have the bandwidth right now to refactor it, so downstream users who want to make use of it will need to watch out for gotchas like this.

Having the information available about how much breaks, and how much attention would be needed, if the coercion was removed would be useful, but no guarantee of progress on this issue on the <3 months timescale!
